agent/tasks/executetestrun.go

```diff
@@ -50,43 +60,37 @@ func (ett ExecuteTestRunTask) Run() error {
 		return errors.New("Problem retrieving testrun from agent cache")
 	}
 
-	// Generate path to testlet and make sure it exists.
-	testlet_path := fmt.Sprintf("%s/assets/testlets/%s", ett.Config.LocalResources.OptDir, tr.Testlet)
-	if _, err := os.Stat(testlet_path); os.IsNotExist(err) {
-		log.Errorf("Testlet %s does not exist on this agent", tr.Testlet)
-		return errors.New("Error executing testrun - testlet doesn't exist on this agent.")
-	}
-
 	log.Debugf("IMMA FIRIN MAH LAZER (for test %s) ", ett.TestUuid)
 
-	// Use a wait group to ensure that all of the testlets have a chance to finish
-	var wg sync.WaitGroup
+	// Specify size of wait group equal to number of targets
 	wg.Add(len(tr.Targets))
 
-	// gatheredData represents test data from this agent for all targets.
-	// Key is target name, value is JSON output from testlet for that target
-	gatheredData := make(map[string]string)
+	testletPath, err := testing.GetTestletPath(tr.Testlet, ett.Config.LocalResources.OptDir)
+	if err != nil {
+		return err
+	}
 
 	// Execute testlets against all targets asynchronously
 	for i := range tr.Targets {
 
 		thisTarget := tr.Targets[i]
 
 		go func() {
+
 			defer wg.Done()
 
-			log.Debugf("Full testlet command and args: '%s %s %s'", testlet_path, thisTarget, tr.Args)
-			cmd := exec.Command(testlet_path, thisTarget, tr.Args)
+			log.Debugf("Full testlet command and args: '%s %s %s'", testletPath, thisTarget, tr.Args)
+			cmd := exec.Command(testletPath, thisTarget, tr.Args)
```

Comment:
General note, this could be changed to use `exec.CommandContext` (added in 1.7) and simplify the timeout logic below.

``` go
ctx, cancel := context.WithTimeout(context.Background(), time.Duration(ett.TimeLimit) * time.Second)
defer cancel()
cmd := exec.CommandContext(ctx, testletPath, thisTarget, tr.Args)
```